sdks/python/apache_beam/io/gcp/bigquery.py

```diff
@@ -661,8 +667,18 @@ def __init__(self, source, test_bigquery_client=None, use_legacy_sql=True,
 
   def _get_source_table_location(self):
     tr = self.source.table_reference
-    if tr is None:
-      # TODO: implement location retrieval for query sources
+    if tr is None:  # it's a query source
+      if self.source.location is not None:
+        logging.info('Using location=%r from BigQuerySource',
+                     self.source.location)
+        return self.source.location
+
+      # TODO(BEAM-5457): Is there a way to get the source location from a query?
```

Comment:
I've investigated this in the past on another PR, and you can run a query in dry-run mode to get the list of referenced tables: https://github.com/apache/beam/pull/5435#discussion_r213817094

Reply:
I ended up implementing it: https://github.com/apache/beam/pull/6463/commits/334b1050d110a8f2635767976888d16b3eeee467

Reply:
IMO, TODO(<bug number>) means it was meant to be included in the scope of the bug.